### DISCUSSION TITLE
How to import local modules in next.config.js

### DISCUSSION THREAD

**Discussion by gzhhong** [vercel/next.js#43017] [Help] [answered]
Hello, 

I want to define all of the rewrites and redirects rules in file ./utils/RouteConfig.js as below:

./utils/RouteConfig.js
```
const RewriteRules = [
    { source: '/fr-FR/explorer', destination: '/fr-FR/explore', locale: false },
]

const RedirectRules = [
    { source: '/fr-FR/explore', destination: "/fr-FR/explorer", locale: false, permanent: true },
]

module.exports = { RewriteRules, RedirectRules };
```
next.config.js
```
const {RewriteRules, RedirectRules} = require('./utils/RouteConfig.js');

const rewrites = () => {
    return RewriteRules;
}
const redirects = () => {
    return RedirectRules;
}
// and other code here
```

The error logs:
```
next start -p 8080
warning Cannot find a suitable global folder. Tried these: "/usr/local, /.yarn"
ready - started server on 0.0.0.0:8080, url: http://localhost:8080
error - Failed to load next.config.js, see more info here https://nextjs.org/docs/messages/next-config-error
Error: Cannot find module './utils/RouteConfig'
Require stack:
- /app/next.config.js
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:985:15)
    at Function.mod._resolveFilename (/app/node_modules/next/dist/build/webpack/require-hook.js:183:28)
    at Function.Module._load (node:internal/modules/cjs/loader:833:27)
    at Module.require (node:internal/modules/cjs/loader:1057:19)
    at require (node:internal/modules/cjs/helpers:103:18)
    at Object.<anonymous> (/app/next.config.js:2:39)
    at Module._compile (node:internal/modules/cjs/loader:1155:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1209:10)
    at Module.load (node:internal/modules/cjs/loader:1033:32)
    at Function.Module._load (node:internal/modules/cjs/loader:868:12) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [ '/app/next.config.js' ]
}
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```
Any suggestions?

Thanks a lot,

James

**Comment by icyJoseph** [comment, marked as answer]
Weird, I have used this in various places. 

```js
const { RedirectRules } = require("./utils/redirects");

/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: true,
  swcMinify: true,
  async redirects() {
    return RedirectRules;
  },
};

module.exports = nextConfig;
```

Never mind the type annotations.

Are you using Docker? I wonder if you are not copying the `utils` folder from the Builder  into your Runner, where `app` is executed. Are you using `standalone` mode?

## Solution

OP used Docker, with a Builder + Runner pattern. OP just needed to copy the `utils/file.js` into the Runner environment.

**Reply by gzhhong** [reply]
Hello icyJoseph,

Yes I am using docker.  I checked the docker file and I didn't find I copy this file from build to runner. I will try as your suggested.

Thank you so much,

James

**Reply by gzhhong** [reply]
Hello icyJoseph,

The problem fixed after I copy this file from docker builder to runner. Thanks a lot for the help.

One more question, I have other files in the utils folder, they are used by code in /pages (but not by next.config.js), why there is no error during running even I didn't copy those files? 

James

**Reply by icyJoseph** [reply]
When building the `.next` folder, the compiler walks the from pages files, downwards, so everything that's touched gets bundled up.

`next.config.js` is not part of the compiling process, at least not in normal mode, I think standalone mode does touch it, and even serializes the config into the standalone output.

You could even build the page with one `next.config.js` and run it with another configuration, of course that doesn't seem to serve any practical use, but it is possible. You really just need the config file if you've made changed the defaults.

It is also the reason why when you modify `next.config.js`, you gotta restart the server. 

Please don't forget to mark as answered.

**Reply by gzhhong** [reply]
@icyJoseph , Hello icyJoseph, Thanks a lot for the help!
